### PR TITLE
Added ability to get metainformation of executive objects used in scr…

### PR DESCRIPTION
**Pull request by kbatoev**
…iptEngineWorker

### CODE REVIEW

**Comment by iakov** [comment]
Let's meet to discuss and review

**Comment by kbatoev** [comment]
Ok

File: trikScriptRunner/include/trikScriptRunner/trikScriptRunner.h [hunk 1]
```diff
@@ -55,6 +55,10 @@ class TrikScriptRunner : public QObject
 	/// Adds custom initialization steps when creating script engine (useful when used from outside of the TrikRuntime).
 	void addCustomEngineInitStep(const std::function<void (QScriptEngine *)> &step);
 
+	/// Gets all methods from executive objects(brick, script, etc) from ScriptEngineWorker
+	/// (useful when used from outside of the TrikRuntime).
+	QStringList getQTSMethodsForLanguage() const;
```
Comment (kbatoev) [inline trikScriptRunner/include/trikScriptRunner/trikScriptRunner.h]: allKnownKeyWords?

File: trikScriptRunner/src/scriptEngineWorker.cpp [hunk 2]
```diff
@@ -48,30 +51,43 @@ using namespace trikScriptRunner;
...
-Q_DECLARE_METATYPE(MotorInterface*)
-Q_DECLARE_METATYPE(ObjectSensorInterface*)
-Q_DECLARE_METATYPE(SoundSensorInterface*)
-Q_DECLARE_METATYPE(SensorInterface*)
-Q_DECLARE_METATYPE(Threading*)
-Q_DECLARE_METATYPE(VectorSensorInterface*)
 Q_DECLARE_METATYPE(QVector<int>)
 Q_DECLARE_METATYPE(trikKernel::TimeVal)
-Q_DECLARE_METATYPE(QTimer*)
+
```
Comment (kbatoev) [inline trikScriptRunner/src/scriptEngineWorker.cpp]: write comment for this (reason, etc.)

File: trikScriptRunner/src/scriptEngineWorker.cpp [hunk 3]
```diff
@@ -48,31 +49,48 @@ using namespace trikScriptRunner;
...
+#define DECLARE_METATYPE_TEMPLATE(TYPE) \
+	Q_DECLARE_METATYPE(TYPE*)
+
+#define REGISTER_METATYPE_FOR_ENGINE(TYPE) \
+	Scriptable<TYPE>::registerMetatype(engine);
+
+#define REGISTER_METATYPE(TYPE) \
+	qRegisterMetaType<TYPE*>(TYPE::staticMetaObject.className());
+
+/// Here we define a comfortablle template that registers all devices used in trik.
```
Comment (dvvrd) [inline trikScriptRunner/src/scriptEngineWorker.cpp]: 'comfortable' is an adjective for my chair. Here you probably mean 'convenient' or 'reusable'. As for me, this whole comment should be like "ATTENTION: do not forget to append newly created device to this list!"

File: trikScriptRunner/include/trikScriptRunner/trikScriptRunner.h [hunk 4]
```diff
@@ -55,6 +55,10 @@ class TrikScriptRunner : public QObject
 	/// Adds custom initialization steps when creating script engine (useful when used from outside of the TrikRuntime).
 	void addCustomEngineInitStep(const std::function<void (QScriptEngine *)> &step);
 
+	/// Gets all methods from executive objects(brick, script, etc) from ScriptEngineWorker
+	/// (useful when used from outside of the TrikRuntime).
+	QStringList getAllKnownMethods() const;
```
Comment (dvvrd) [inline trikScriptRunner/include/trikScriptRunner/trikScriptRunner.h]: Bad name. As for me `knownKeywords` sounds better.

File: trikScriptRunner/src/scriptEngineWorker.h [hunk 5]
```diff
@@ -133,6 +136,11 @@ private slots:
 	/// Evaluates "system.js" file in given engine.
 	void evalSystemJs(QScriptEngine * const engine) const;
 
+	/// Collects all methods names from given metaObject.
+	/// If the returnType of given method name is registered in metaobject system,
+	/// newMetaObject is constructed and procedure is called recursively for it.
+	void collectMethods(QSet<QString> &result, const QMetaObject *obj) const;
```
Comment (dvvrd) [inline trikScriptRunner/src/scriptEngineWorker.h]: `collectMethodNames` is a better name

File: trikScriptRunner/src/scriptEngineWorker.h [hunk 6]
```diff
@@ -73,6 +73,9 @@ class ScriptEngineWorker : public QObject
 	/// Can be safely called from other threads.
 	void stopScript();
 
+	/// Collects all methods and root objects names from all trees.
+	QStringList getAllKnownMethods() const;
```
Comment (dvvrd) [inline trikScriptRunner/src/scriptEngineWorker.h]: `knownMethodNames`

File: trikScriptRunner/src/scriptEngineWorker.cpp [hunk 7]
```diff
@@ -386,3 +401,23 @@ void ScriptEngineWorker::evalSystemJs(QScriptEngine * const engine) const
 		engine->globalObject().setProperty(functionName, functionValue);
 	}
 }
+
+void ScriptEngineWorker::collectMethods(QSet<QString> &result, const QMetaObject *obj) const
+{
+	for (int i = obj->methodOffset(); i < obj->methodCount(); ++i) {
+		QMetaMethod metaMethod = obj->method(i);
+		QString methodName = QString::fromLatin1(metaMethod.name());
```
Comment (dvvrd) [inline trikScriptRunner/src/scriptEngineWorker.cpp]: Within our styleguide `const` should be used in all possible places.

File: trikScriptRunner/src/scriptEngineWorker.cpp [hunk 8]
```diff
@@ -386,3 +401,23 @@ void ScriptEngineWorker::evalSystemJs(QScriptEngine * const engine) const
...
+		result.insert(methodName);
+
+		QString methodReturnType = QString::fromLatin1(metaMethod.typeName());
+		if (methodReturnType.endsWith('*')) {
+			methodReturnType.chop(1);
+		}
+
+		int typeId = QMetaType::type(methodReturnType.toLatin1());
```
Comment (dvvrd) [inline trikScriptRunner/src/scriptEngineWorker.cpp]: `const`

File: trikScriptRunner/include/trikScriptRunner/trikScriptRunner.h [hunk 9]
```diff
@@ -55,6 +55,10 @@ class TrikScriptRunner : public QObject
 	/// Adds custom initialization steps when creating script engine (useful when used from outside of the TrikRuntime).
 	void addCustomEngineInitStep(const std::function<void (QScriptEngine *)> &step);
 
+	/// Gets all methods from executive objects(brick, script, etc) from ScriptEngineWorker
```
Comment (dvvrd) [inline trikScriptRunner/include/trikScriptRunner/trikScriptRunner.h]: Actually it returns method names, not methods themselves. Missed whitespace near '('. Missed '.' after 'etc'.

**Pull request merged**